catroid/src/androidTest/java/org/catrobat/catroid/test/embroidery/DSTFileGeneratorTest.java

```diff
@@ -0,0 +1,214 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertTrue;
+
+@RunWith(AndroidJUnit4.class)
+public class DSTFileGeneratorTest {
+	final String projectName = "testProject";
+	DSTFileGenerator fileGenerator;
+	File dstFile;
+	final int[] conversion = {-81, 81, -27, 27, -9, 9, -3, 3, -1, 1};
+
+	@Before
+	public void setUp() {
+		Project project = new Project(InstrumentationRegistry.getTargetContext(), projectName);
+		ProjectManager.getInstance().setCurrentProject(project);
+		fileGenerator = new DSTFileGenerator();
+		dstFile = new File(Constants.CACHE_DIR + "/" + projectName + ".dst");
+	}
+
+	private int getValueForByte(int element) {
+		int mask = 0x200;
+		int value = 0;
+		for (int i = 0; i < conversion.length; i++) {
+			if ((element & (mask >>> i)) != 0) {
+				value += conversion[i];
+			}
+		}
+		return value;
+	}
+
+	@Test
+	public void testConversionTable() {
+		for (int element = 0; element < DSTFileGenerator.CONVERSIONTABLE.length; element++) {
+			int value = getValueForByte(DSTFileGenerator.CONVERSIONTABLE[element]);
+			if (element > 121) {
+				assertEquals((element - 121) * (-1), value);
+			} else {
+				assertEquals(element, value);
+			}
+		}
+	}
+
+	@Test
+	public void testWriteToSampleDSTFile() throws IOException, URISyntaxException {
+		EmbroideryList stitchpoints = new EmbroideryList();
+		stitchpoints.add(new PointF(-100, 0));
+		stitchpoints.add(new PointF(100, 0));
+		stitchpoints.add(new PointF(100, 100));
+		stitchpoints.add(new PointF(0, 150));
+		stitchpoints.add(new PointF(-100, 100));
+		stitchpoints.add(new PointF(-100, 0));
+		stitchpoints.add(new PointF(100, 100));
+		stitchpoints.add(new PointF(-100, 100));
+		stitchpoints.add(new PointF(100, 0));
+		if (dstFile.exists()) {
+			dstFile.delete();
+		}
+		dstFile.createNewFile();
+		fileGenerator.writeToDSTFile(dstFile, stitchpoints);
+
+		InputStream inputStream = InstrumentationRegistry.getContext().getResources().openRawResource(org.catrobat
+				.catroid.test.R.raw.sample_dst_file);
+		File compareFile = StorageOperations.copyStreamToDir(inputStream, Constants.CACHE_DIR, "sample_dst_file.dst");
+
+		assertEquals(compareFile.length(), dstFile.length());
+
+		byte[] compareFileBytes = FileUtils.readFile(compareFile);
+		byte[] dstFileBytes = FileUtils.readFile(dstFile);
+
+		boolean fileBytesEquality = Arrays.equals(compareFileBytes, dstFileBytes);
+		assertTrue(fileBytesEquality);
+	}
+
+	@Test
+	public void testPrepareStitchPoints() {
+		final int expectedAmountJumpPoints = 2;
+		EmbroideryList points = new EmbroideryList();
+		points.add(new PointF(0, 0));
+		points.add(new PointF(1000, 1000));
+		ArrayList<DSTFileGenerator.StitchPoint> stitchPoints = fileGenerator.prepareStitchPoints(points);
+
+		assertEquals(points.size() + expectedAmountJumpPoints, stitchPoints.size());
+
+		int index = 0;
+		for (int i = 0; i < stitchPoints.size(); i++) {
```

Comment:
loops and branching statements should be avoided in tests, this double index loop construction is extremely hard to read and comprehend.
why not construct a EmbroideryList points in a way that it will end up having a reasonable size after prepareStitchPoints and have an expected ArrayList<DSTFileGenerator.StitchPoint> that you compare it to.

or construct a EmbroideryListPoints points so small that you make sure that it interpolates just one point that you can assert against. and the same with jump points.